File: hometask_8/Source_8.3/list.cpp
#include "List.h"

ListElement *create(WordInfo *word) {
	ListElement *newEl = new ListElement;
	newEl->word = word;
	newEl->next = nullptr;

	return newEl;
}

bool isWordIn(String text, List &list) {
	for (int i = 0; i < list.size; ++i) {
		WordInfo curWord = get(i, list);
		if (areEqual(text, curWord.text)) {
			setNumber(i, list, curWord.number + 1);
			return true;
		}
	}
	return false;
}

void add(WordInfo *word, List &list) {
	if (isEmpty(list)) {
		list.first = create(word);
		list.last = list.first;
		list.size++;
	}
	else {
		if (!isWordIn(word->text, list)) {
			list.last->next = create(word);
			list.last = list.last->next;
			list.size++;
		}
		else
			delete(word);
	}
}

WordInfo get(int i, List &list) {
	ListElement *temp = list.first;
	while (i > 0) {
		temp = temp->next;
		--i;
	}

	return *temp->word;
}

void setNumber(int i, List &list, int number) {
	ListElement *temp = list.first;
	while (i > 0) {
		temp = temp->next;
		--i;
	}
	temp->word->number = number;
}

void remove(int i, List &list) {
	if (i == 0) {
		deleteWord(list.first->word);
		list.first = list.first->next;
	}
	else {
		ListElement *temp = list.first;
		while (i > 1) {
			temp = temp->next;
			--i;
		}

		deleteWord(temp->next->word);
		temp->next = temp->next->next;
	}
	list.size--;
}

bool isEmpty(List &list) {
	return list.first == nullptr;
}

void clear(List &list) {
	while (!isEmpty(list))
		remove(0, list);
}

int getFrequencyOfWord(String text, List &list) {
	for (int i = 0; i < list.size; ++i) {
		WordInfo curWord = get(i, list);
		if (areEqual(text, curWord.text))
			return curWord.number;
	}
	return 0;
}

int numberOfWords(List &list) {
	int result = 0;
	for (int i = 0; i < list.size; ++i) {
		result += get(i, list).number;
	}

	return result;
}

File: semester_2/hometask_3/Source_3.1/src/main/java/group144/khodko/List.java
package group144.khodko;

public class List {
    private class ListElement {
        String text;
        int frequency = 1;
        ListElement next;

        ListElement(String text) {
            this.text = text;
        }
    }

    private ListElement first;
    private ListElement last;
    private int size = 0;


    public void add(String word) {
        int i = 0;
        ListElement current = first;
        while (i < size) {
            if (current.text.equals(word)) {
                current.frequency++;
                return;
            }
            current = current.next;
            i++;
        }

        if (isEmpty()) {
            first = new ListElement(word);
            last = first;
        }
        else {
            last.next = new ListElement(word);
            last = last.next;
        }
        size++;
    }

    public void remove(int i) {
        if (i >= size || i < 0)
            return;

        ListElement current = first;
        ListElement previous = first;
        while (i != 0) {
            previous = current;
            current = current.next;
            i--;
        }
        previous.next = current.next;
        size--;
    }

    public String get(int i) {
        if (i >= size || i < 0)
            throw new IndexOutOfBoundsException();

        ListElement current = first;
        while (i != 0) {
            current = current.next;
            i--;
        }
       return current.text;
    }

    public int getFrequencyOfWord(String word) {
        ListElement current = first;

        int i = 0;
        while (i < size) {
            if (current.text.equals(word))
                return current.frequency;
            current = current.next;
            i++;
        }
        return 0;
    }

    public int numberOfWords() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }
}
